misk-hibernate/src/main/kotlin/misk/hibernate/SchemaMigrator.kt

```diff
@@ -23,31 +20,105 @@ private val logger = getLogger<SchemaMigrator>()
  *
  * Available schema migrations are SQL files in the datasource's `migrations_resource` directory.
+ * Multiple directories can be specified in `migrations_resources` to support dependencies with
+ * their own migrations.
+ *
  * Each file should contain SQL statements terminated by a `;`. The files should be named like
  * `v100__exemplar.sql` with a `v`, an integer version, two underscores, a description, and the
  * `.sql` suffix. The integer identifier is the migration version. Versions do not need to be
  * sequential. They are applied in increasing order.
  *
+ * Migrations found in subdirectories will become namespaced, where the namespace is the directory
+ * path. This allows including migrations from dependencies. Migrations in the root directory have
+ * versions "1", "2", etc. Namespaced migrations have versions in the form of "dir/1", "dir/sub/1"
+ * 
  * Applied schema migrations are tracked in the database in a `schema_version` table. Migrations may
  * be applied either by [SchemaMigrator.applyAll] or manually. When you applying schema changes
  * manually you must add a row to the `schema_version` table to record which version was applied.
  */
+
+data class NamedspacedMigration(
+  val namespace: String,
+  val version: Int
+) : Comparable<NamedspacedMigration> {
+  // We don't store the path from which schema changes came, so we don't use it for comparison
+  // between what's in the database and what's available.
+  var path = ""
+
+  constructor(version: Int) : this("", version)
+
+  override fun compareTo(other: NamedspacedMigration): Int {
+    if (namespace == other.namespace) {
+      return version.compareTo(other.version)
+    } else {
+      return namespace.compareTo(other.namespace)
+    }
+  }
+
+  fun toNamespacedVersion() = "$namespace$version"
+
+  companion object {
+    fun fromNamespacedVersion(namespacedVersion: String): NamedspacedMigration {
+      if (namespacedVersion.toIntOrNull() != null) {
+        return NamedspacedMigration(namespacedVersion.toInt())
+      } else {
+        val items = namespacedVersion.split("/")
+        val version = items.last().toInt()
+        val namespace = items.dropLast(1).joinToString("/") + "/"
+        return NamedspacedMigration(namespace, version)
+      }
+    }
+
+    fun fromResourcePath(resource: String, migrationsResource: String): NamedspacedMigration {
+      val matcher = MIGRATION_PATTERN.matcher(resource)
+      require(matcher.matches()) { "unexpected resource: $resource" }
+      val cleanNamespace = matcher.group(1).removePrefix(migrationsResource).removePrefix("/")
+      val namedspacedMigration = NamedspacedMigration(cleanNamespace, matcher.group(2).toInt())
+      namedspacedMigration.path = resource
+      return namedspacedMigration
+    }
+
+    /** Matches file names like `exemplar/migrations/v100__exemplar.sql`. */
+    val MIGRATION_PATTERN = Pattern.compile("(.*/?)v(\\d+)__[^/]+\\.sql")!!
```

Comment:
I think you need a slash preceding the `v`

I think this matches: `foo/luv1__franklin.sql`

Reply:
... looks like this bug was already here (my fault!)

Reply:
I'll add a test case for it